### PR TITLE
Keyboard nav, context menu support

### PR DESCRIPTION
**Pull request by Blackbaud-ErikaMcVey**
Changes include:
- sky ag grid wrapper component
- keyboard navigation
- popovers and datepicker updates
- sticky column headers via viewkeeper
- allowing the checkbox component to not be set from provided data
- context menu example

### CODE REVIEW

File: src/app/public/modules/ag-grid/index.ts [hunk 1]
```diff
@@ -1,3 +1,5 @@
+export * from './ag-grid.service';
```
Comment (Blackbaud-AlexKingman) [inline src/app/public/modules/ag-grid/index.ts:1]: Looks like you've got the service in here twice?

File: src/app/public/modules/ag-grid/index.ts [hunk 2]
```diff
@@ -1,3 +1,5 @@
...
 export * from './ag-grid.module';
 export * from './ag-grid.service';
 export * from './types';
+export * from './ag-grid-wrapper.component';
```
Comment (Blackbaud-AlexKingman) [inline src/app/public/modules/ag-grid/index.ts:5]: Does the consumer ever need access to the component? If not, we probably don't need to export it.

File: src/app/visual/readonly-grid/readonly-grid-context-menu.component.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+@Component({
+  selector: 'readonly-grid-context-menu',
+  templateUrl: './readonly-grid-context-menu.component.html',
+  changeDetection: ChangeDetectionStrategy.OnPush
+})
+export class ReadonlyGridContextMenuComponent implements ICellRendererAngularComp {
+  public deleteLabel: string;
+  public markInactiveLabel: string;
+  public moreInfoLabel: string;
+  public goalName: string;
```
Comment (Blackbaud-AlexKingman) [inline src/app/visual/readonly-grid/readonly-grid-context-menu.component.ts:23]: Check your alphabetical ordering here.

File: src/app/visual/readonly-grid/readonly-grid.component.ts [hunk 4]
```diff
@@ -6,14 +6,19 @@ import {
 import {
   GridReadyEvent,
   GridOptions,
-  ICellRendererParams
+  ICellRendererParams,
+  GridApi
```
Comment (Blackbaud-AlexKingman) [inline src/app/visual/readonly-grid/readonly-grid.component.ts:10]: Check your alphabetical ordering here.

File: src/app/visual/readonly-grid/readonly-grid.component.ts [hunk 5]
```diff
@@ -35,6 +40,7 @@ export class ReadonlyGridComponent implements OnInit {
   public gridData = READONLY_GRID_DATA;
   public hasMore = true;
   public gridOptions: GridOptions;
+  public gridApi: GridApi;
```
Comment (Blackbaud-AlexKingman) [inline src/app/visual/readonly-grid/readonly-grid.component.ts:43]: Alphabetical ordering of properties.

File: src/assets/locales/resources_en_US.json [hunk 6]
```diff
@@ -14,5 +14,17 @@
   "sky_ag_grid_cell_editor_autocomplete_aria_label": {
     "_description": "aria label for the autocomplete cell editor",
     "message": "Editable autocomplete {0} for row {1}"
+  },
+  "readonly_grid_delete_text": {
+    "_description": "Context menu delete text for readonly grid example",
```
Comment (Blackbaud-AlexKingman) [inline src/assets/locales/resources_en_US.json:19]: Do we really need resource strings for visual tests?

File: src/app/public/modules/ag-grid/fixtures/ag-grid-wrapper-adapter.component.fixture.ts [hunk 7]
```diff
@@ -0,0 +1,10 @@
+
```
Comment (Blackbaud-AlexKingman) [inline src/app/public/modules/ag-grid/fixtures/ag-grid-wrapper-adapter.component.fixture.ts:1]: Looks like an extra line here?

File: src/app/visual/readonly-grid/readonly-grid-context-menu.component.html [hunk 8]
```diff
@@ -0,0 +1,30 @@
+<sky-dropdown
+  buttonType="context-menu"
+>
+  <sky-dropdown-menu>
+    <sky-dropdown-item>
+      <button
```
Comment (Blackbaud-AlexKingman) [inline src/app/visual/readonly-grid/readonly-grid-context-menu.component.html:6]: If I navigate to a dropdown with the keyboard controls, open the dropdown, then hit `tab`, the focus is placed on the next dropdown button. Is that expected?

File: src/app/visual/readonly-grid/readonly-grid.component.html [hunk 9]
```diff
@@ -1,11 +1,13 @@
 <div
   class="readonly-grid"
 >
-  <ag-grid-angular
-    [gridOptions]="gridOptions"
-    [rowData]="gridData"
-  >
-  </ag-grid-angular>
+  <sky-ag-grid-wrapper>
```
Comment (Blackbaud-AlexKingman) [inline src/app/visual/readonly-grid/readonly-grid.component.html:4]: Maybe not part of this PR, but I noticed there's no way to sort the columns with keyboard controls? Maybe I'm missing it.

File: src/app/public/modules/ag-grid/cell-editors/cell-editor-datepicker/cell-editor-datepicker.component.html [hunk 10]
```diff
@@ -5,7 +5,6 @@
 >
   <sky-datepicker
```
Comment (Blackbaud-AlexKingman) [inline src/app/public/modules/ag-grid/cell-editors/cell-editor-datepicker/cell-editor-datepicker.component.html:6]: The datepicker button can only be opened with `space`, but not the `enter` key. Is that expected?

File: src/app/visual/editable-grid/editable-grid.component.html [hunk 11]
```diff
@@ -1,37 +1,39 @@
...
-  <ag-grid-angular
-    [ngClass]="{'sky-ag-grid-editable': editMode}"
-    [gridOptions]="gridOptions"
-    [rowData]="gridData"
-  >
-  </ag-grid-angular>
+  <sky-ag-grid-wrapper>
+    <sky-toolbar>
+      <sky-toolbar-item *ngIf="!editMode">
+        <button
```
Comment (Blackbaud-AlexKingman) [inline src/app/visual/editable-grid/editable-grid.component.html:7]: Now that this button is inside the wrapper, I can't navigate to it with the `tab` key.

File: src/app/public/modules/ag-grid/ag-grid-wrapper-adapter.service.ts [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+    if (element) {
+      element.focus();
+    }
+  }
+
+  public getFocusedElement(): HTMLElement {
+    return document.activeElement as HTMLElement;
+  }
+
+  public focusOnFocusableChildren(element: HTMLElement): void {
```
[the remark this answers is not in the document]
Comment (Blackbaud-ErikaMcVey) [reply src/app/public/modules/ag-grid/ag-grid-wrapper-adapter.service.ts:37]: So that takes an `ElementRef` which I don't necessarily have. Is there an easy way to convert a regular old `Element` to an `ElementRef`?

File: src/app/public/modules/ag-grid/ag-grid-wrapper-adapter.service.ts [hunk 13]
```diff
@@ -0,0 +1,44 @@
+import {
+  Injectable
+} from '@angular/core';
+
+import {
+  SkyCoreAdapterService
+} from '@skyux/core';
+
+@Injectable()
+export class SkyAgGridWrapperAdapterService {
```
[the remark this answers is not in the document]
Comment (Blackbaud-ErikaMcVey) [reply src/app/public/modules/ag-grid/ag-grid-wrapper-adapter.service.ts:10]: ah good point, I'll rename it. The service that configures the grid options actually already uses it for something that isn't specific to the wrapper element.

File: src/assets/locales/resources_en_US.json (hunk 6 above)
Comment (Blackbaud-ErikaMcVey) [reply src/assets/locales/resources_en_US.json:19]: ah, this was originally to make sure localization still worked for a different approach to the context menu. Since it's just a standard component now this isn't really necessary.

File: src/app/visual/readonly-grid/readonly-grid.component.html (hunk 9 above)
Comment (Blackbaud-ErikaMcVey) [reply src/app/visual/readonly-grid/readonly-grid.component.html:4]: hmm I'll look into this, looks like ag-grid doesn't let you focus on the column headers 🤔

File: src/app/visual/readonly-grid/readonly-grid-context-menu.component.html (hunk 8 above)
Comment (Blackbaud-ErikaMcVey) [reply src/app/visual/readonly-grid/readonly-grid-context-menu.component.html:6]: this was something we talked about in flex as not being the most ideal but something we could put up with. We still have to allow the tab key in popups/editors because a lot of sky components use tab to navigate them, like the datepicker. Without modifying each sky component I couldn't really find an easy way around it. Once they hit tab again or use the arrow keys it resumes the expected focus/highlight.

File: src/app/visual/editable-grid/editable-grid.component.html (hunk 11 above)
Comment (Blackbaud-ErikaMcVey) [reply src/app/visual/editable-grid/editable-grid.component.html]: good catch. I initially was trying to get the toolbar to be sticky in this PR, but after discussing with Paul I think that functionality belongs in data manager. This way the sky ag grid wrapper can focus on _just_ the grid.

File: src/app/public/modules/ag-grid/cell-editors/cell-editor-datepicker/cell-editor-datepicker.component.html (hunk 10 above)
Comment (Blackbaud-ErikaMcVey) [reply src/app/public/modules/ag-grid/cell-editors/cell-editor-datepicker/cell-editor-datepicker.component.html:6]: kind of? Hitting `enter` is the only way to either start editing a cell or save the edits made via the keyboard. I could try to suppress the event from the datepicker editor when the button is focused if we think it is important to be able to use `enter` or `space`. My initial reason for not doing so was just trying to avoid complicating keyboard interactions even more.

Comment (Blackbaud-AlexKingman) [reply src/app/public/modules/ag-grid/cell-editors/cell-editor-datepicker/cell-editor-datepicker.component.html:6]: I'm not sure of the right answer here. We might have to leave this one up to an accessibility audit. I'm going to resolve this thread as its not a dealbreaker, IMO.

File: src/app/visual/readonly-grid/readonly-grid-context-menu.component.html (hunk 8 above)
Comment (Blackbaud-AlexKingman) [reply src/app/visual/readonly-grid/readonly-grid-context-menu.component.html:6]: Got it. Thanks for the explanation.

File: src/app/public/modules/ag-grid/ag-grid-wrapper-adapter.service.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (Blackbaud-AlexKingman) [reply src/app/public/modules/ag-grid/ag-grid-wrapper-adapter.service.ts]: Hmm... looking at how you need to grab the element with focus, I'm not sure there's an easy way to get the `ElementRef` on the fly like that. Let's pass on this for now.
